### ISSUE TITLE
[Site Bug] SQL syntax error [WAS: Issue with the page: /timescaledb/latest/how-to-guides/user-defined-actions/example-tiered-storage/]

### ISSUE THREAD

**Issue by jouir** [timescale/docs#323]
# Describe the bug

There is a syntax error in the `move_chunks` procedure:

```
ERROR:  missing "LOOP" at end of SQL expression
LINE 25:   WHERE pgts.spcname != destination;
```

## What do the docs say now?

```
CREATE OR REPLACE PROCEDURE move_chunks (job_id int, config jsonb)
LANGUAGE PLPGSQL
AS $$
DECLARE
  ht REGCLASS;
  lag interval;
  destination name;
  chunk REGCLASS;
  tmp_name name;
BEGIN
  SELECT jsonb_object_field_text (config, 'hypertable')::regclass INTO STRICT ht;
  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
  SELECT jsonb_object_field_text (config, 'tablespace') INTO STRICT destination;

  IF ht IS NULL OR lag IS NULL OR destination IS NULL THEN
    RAISE EXCEPTION 'Config must have hypertable, lag and destination';
  END IF;

  FOR chunk IN
  SELECT show.oid
  FROM show_chunks(ht, older_than => lag)
  SHOW (oid)
    INNER JOIN pg_class pgc ON pgc.oid = show.oid
    INNER JOIN pg_tablespace pgts ON pgts.oid = pgc.reltablespace
  WHERE pgts.spcname != destination;
  LOOP
    RAISE NOTICE 'Moving chunk: %', chunk::text;
    EXECUTE format('ALTER TABLE %s SET TABLESPACE %I;', chunk, destination);
  END LOOP;
END
$$;
```

## What should the docs say?

```
CREATE OR REPLACE PROCEDURE move_chunks (job_id int, config jsonb)
LANGUAGE PLPGSQL
AS $$
DECLARE
  ht REGCLASS;
  lag interval;
  destination name;
  chunk REGCLASS;
  tmp_name name;
BEGIN
  SELECT jsonb_object_field_text (config, 'hypertable')::regclass INTO STRICT ht;
  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
  SELECT jsonb_object_field_text (config, 'tablespace') INTO STRICT destination;

  IF ht IS NULL OR lag IS NULL OR destination IS NULL THEN
    RAISE EXCEPTION 'Config must have hypertable, lag and destination';
  END IF;

  FOR chunk IN
  SELECT show.oid
  FROM show_chunks(ht, older_than => lag)
  SHOW (oid)
    INNER JOIN pg_class pgc ON pgc.oid = show.oid
    INNER JOIN pg_tablespace pgts ON pgts.oid = pgc.reltablespace
  WHERE pgts.spcname != destination
  LOOP
    RAISE NOTICE 'Moving chunk: %', chunk::text;
    EXECUTE format('ALTER TABLE %s SET TABLESPACE %I;', chunk, destination);
  END LOOP;
END
$$;
```

# Page affected

https://docs.timescale.com/timescaledb/latest/how-to-guides/user-defined-actions/example-tiered-storage/#tiered-storage

# Version affected

latest

# Subject matter expert (SME)

# Screenshots

![Screenshot from 2021-08-31 12-14-27](https://user-images.githubusercontent.com/24910974/131485249-b4cd5b98-54dc-4c3d-999f-797544dbe5f6.png)

# Any further info

The procedure has been created on:
* PostgreSQL 13.4 (from pgdg)
* Timescale 2.4.1 (from Timescale Debian packaging)

**Comment by jouir** [comment]
Note that the following query doesn't return chunks that are subject to be moved:
```sql
  SELECT show.oid
  FROM show_chunks(ht, older_than => lag)
  SHOW (oid)
    INNER JOIN pg_class pgc ON pgc.oid = show.oid
    INNER JOIN pg_tablespace pgts ON pgts.oid = pgc.reltablespace
  WHERE pgts.spcname != destination;
```
The "oid" here is not an oid (eg: `123`) but the full relation name (eg: `_timescaledb_internal._hyper_1_708_chunk`) and the default tablespace is null so the join on the `pg_tablespace` table excludes chunks created on the default tablespace.

Example:
```
tiering=# SELECT show.oid
tiering-# FROM show_chunks('weather_metrics', older_than => interval '10 years')
tiering-# SHOW (oid)
tiering-#   INNER JOIN pg_class pgc ON pgc.oid = show.oid
tiering-#   INNER JOIN pg_tablespace pgts ON pgts.oid = pgc.reltablespace
tiering-# WHERE pgts.spcname != 'history'
tiering-# ORDER BY show.oid DESC
tiering-# LIMIT 10;
 oid 
-----
(0 rows)
```

Instead, if we use the following query:
```sql
SELECT c.oid
  FROM pg_class AS c
    LEFT JOIN pg_tablespace AS t ON (c.reltablespace = t.oid)
    JOIN pg_namespace AS n ON (c.relnamespace = n.oid)
    JOIN (SELECT * FROM show_chunks(ht, older_than => lag) SHOW (oid)) AS chunks ON (chunks.oid::text = n.nspname || '.' || c.relname)
WHERE t.spcname != destination OR t.spcname IS NULL;
```

The chunks are selected:
```
tiering=# SELECT c.oid, n.nspname || '.' || c.relname as chunk, t.spcname as tablespace
  FROM pg_class AS c
    LEFT JOIN pg_tablespace AS t ON (c.reltablespace = t.oid)
    JOIN pg_namespace AS n ON (c.relnamespace = n.oid)
    JOIN (SELECT * FROM show_chunks('weather_metrics', older_than => interval '10 years') SHOW (oid)) AS chunks ON (chunks.oid::text = n.nspname || '.' || c.relname)
WHERE t.spcname != 'history' OR t.spcname IS NULL
LIMIT 10;
  oid  |                  chunk                   | tablespace 
-------+------------------------------------------+------------
 22687 | _timescaledb_internal._hyper_1_714_chunk | (null)
 22695 | _timescaledb_internal._hyper_1_715_chunk | (null)
 22703 | _timescaledb_internal._hyper_1_716_chunk | (null)
 22711 | _timescaledb_internal._hyper_1_717_chunk | (null)
 22719 | _timescaledb_internal._hyper_1_718_chunk | (null)
 22727 | _timescaledb_internal._hyper_1_719_chunk | (null)
 22735 | _timescaledb_internal._hyper_1_720_chunk | (null)
 22743 | _timescaledb_internal._hyper_1_721_chunk | (null)
 22751 | _timescaledb_internal._hyper_1_722_chunk | (null)
 22759 | _timescaledb_internal._hyper_1_723_chunk | (null)
(10 rows)
```

We can even extend the procedure to use the `move_chunk` function as suggested on the documentation:
```sql
CREATE OR REPLACE PROCEDURE move_chunks (job_id int, config jsonb)
LANGUAGE PLPGSQL
AS $$
DECLARE
  ht REGCLASS;
  lag interval;
  destination_tablespace name;
  index_destination_tablespace name;
  reorder_index REGCLASS;
  chunk REGCLASS;
  tmp_name name;
BEGIN
  SELECT jsonb_object_field_text (config, 'hypertable')::regclass INTO STRICT ht;
  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
  SELECT jsonb_object_field_text (config, 'destination_tablespace') INTO STRICT destination_tablespace;
  SELECT jsonb_object_field_text (config, 'index_destination_tablespace') INTO STRICT index_destination_tablespace;
  SELECT jsonb_object_field_text (config, 'reorder_index') INTO STRICT reorder_index;

  IF ht IS NULL OR lag IS NULL OR destination_tablespace IS NULL THEN
    RAISE EXCEPTION 'Config must have hypertable, lag and destination_tablespace';
  END IF;

  IF index_destination_tablespace IS NULL THEN
    index_destination_tablespace := destination_tablespace;
  END IF;

  FOR chunk IN
    SELECT c.oid
    FROM pg_class AS c
      LEFT JOIN pg_tablespace AS t ON (c.reltablespace = t.oid)
      JOIN pg_namespace AS n ON (c.relnamespace = n.oid)
      JOIN (SELECT * FROM show_chunks(ht, older_than => lag) SHOW (oid)) AS chunks ON (chunks.oid::text = n.nspname || '.' || c.relname)
    WHERE t.spcname != destination_tablespace OR t.spcname IS NULL
  LOOP
    RAISE NOTICE 'Moving chunk: %', chunk::text;
    PERFORM move_chunk(
        chunk => chunk,
        destination_tablespace => destination_tablespace,
        index_destination_tablespace => index_destination_tablespace,
        reorder_index => reorder_index
    );
  END LOOP;
END
$$;
```

Arguments have been renamed to be similar to the `move_chunk` function:

```
tiering=# CALL move_chunks(job_id => 1000, config => '{"hypertable":"weather_metrics","lag":"28 year","destination_tablespace":"history","reorder_index":"weather_metrics_time_idx"}');
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_714_chunk
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_715_chunk
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_716_chunk
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_717_chunk
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_718_chunk
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_719_chunk
[...]
NOTICE:  Moving chunk: _timescaledb_internal._hyper_1_766_chunk
CALL
```
And subsequent calls do nothing:
```
tiering=# CALL move_chunks(job_id => 1000, config => '{"hypertable":"weather_metrics","lag":"28 year","destination_tablespace":"history","reorder_index":"weather_metrics_time_idx"}');
CALL
```

**Comment by kfatehi** [comment]
the issue still exists on [the overview page](https://docs.timescale.com/timescaledb/latest/overview/core-concepts/user-defined-actions/#tiered-storage)

**Comment by kfatehi** [comment]
I found this issue by searching the web for the error. I was puzzled because I first had copied the fixed version (for which this issue was closed) and found it was doing nothing. Tried it again using the broken copy of the script (puzzling). I was really happy to see @jouir ’s version because it matched my intuition that the move_chunk function be used in the body of the loop (having prior witnessed that that function does work). So I would suggest that his version be the one published in the doc FWIW.
